### PR TITLE
Snapshot policy resource

### PR DESCRIPTION
**Pull request by AnikaAgiwal2711**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 80% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

**Comment by terraform-providers-svc** [comment]
Can one of the admins verify this patch?

File: docs/resources/snapshot_policy.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
+---
+# Copyright (c) 2023 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shenda1) [inline docs/resources/snapshot_policy.md:2]: 2024

File: examples/resources/powerflex_snapshot_policy/resource.tf [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/*
+Copyright (c) 2023 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shenda1) [inline examples/resources/powerflex_snapshot_policy/resource.tf:2]: 2024

File: powerflex/provider/snapshot_policy_resource_test.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright (c) 2023 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource_test.go:2]: 2024

File: powerflex/provider/snapshot_policy_resource.go [hunk 4]
```diff
@@ -0,0 +1,450 @@
+/*
+Copyright (c) 2023 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:2]: 2024

File: powerflex/provider/snapshot_policy_resource.go [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+	// Assigning volumes to snapshot policy
+	stringListVol := helper.ListToSliceVol(plan)
+	for _, v := range stringListVol {
+		payload2 := &scaleiotypes.AssignVolumeToSnapshotPolicyParam{
+			SourceVolumeId: v,
+		}
+		err2 := r.system.AssignVolumeToSnapshotPolicy(payload2, snapID)
+		if err2 != nil {
+			resp.Diagnostics.AddError(
+				"Error assigning volume to snapshot policy",
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:131]: Since this is a new resource, can we handle the partial creation issue?

File: powerflex/provider/snapshot_policy_resource.go [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+		return
+	}
+
+	stateVolList := helper.ListToSliceVol(state)
+
+	if len(stateVolList) > 0 {
+		for _, v := range stateVolList {
+			payload2 := &scaleiotypes.AssignVolumeToSnapshotPolicyParam{
+				SourceVolumeId:            v,
+				AutoSnapshotRemovalAction: "Detach",
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:420]: Any reason remove mode is hardcoded here?

File: powerflex/provider/snapshot_policy_resource.go [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+	volumes, err := r.system.GetSourceVolume(snapID)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Could not get volumes assigned to snapshot policy"),
+			err.Error(),
+		)
+		return
+	}
+
+	// upadting the list of volumes assinged to the state
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:159]: updating

File: powerflex/provider/snapshot_policy_resource.go [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+	for _, v := range volumes {
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:160]: Since we have a dedicated function for populating state, can this logic be moved to function?

File: powerflex/provider/snapshot_policy_schema.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+				stringvalidator.LengthAtLeast(1),
+			},
+		},
+		"paused": schema.BoolAttribute{
+			Description:         "Indicates that the snapshot policy should paused or not.",
+			MarkdownDescription: "Indicates that the snapshot policy should paused or not.",
+			Optional:            true,
+			Computed:            true,
+			PlanModifiers: []planmodifier.Bool{
+				helper.BoolDefault(false),
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_schema.go:64]: Please check this for setting up the default values https://developer.hashicorp.com/terraform/plugin/framework/resources/default

File: powerflex/provider/snapshot_policy_schema.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+			},
+		},
+		"volume_id": schema.SetAttribute{
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_schema.go:67]: volume_ids?

File: examples/resources/powerflex_snapshot_policy/resource.tf [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+# Command to run this tf file : terraform init && terraform plan && terraform apply
+# Create, Update, Delete is supported for this resource
+# To import , check import.sh for more info
+# name and protection_domain_id is the required parameter to create or update
```
Comment (shenda1) [inline examples/resources/powerflex_snapshot_policy/resource.tf:21]: verify once if protection_domain_id is needed.

File: examples/resources/powerflex_snapshot_policy/resource.tf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+# Command to run this tf file : terraform init && terraform plan && terraform apply
+# Create, Update, Delete is supported for this resource
+# To import , check import.sh for more info
+# name and protection_domain_id is the required parameter to create or update 
+# To check which attributes of the fault set can be updated, please refer Product Guide in the documentation
```
Comment (shenda1) [inline examples/resources/powerflex_snapshot_policy/resource.tf:22]: fault set -> snapshot policy

File: examples/resources/powerflex_snapshot_policy/resource.tf [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+# Command to run this tf file : terraform init && terraform plan && terraform apply
+# Create, Update, Delete is supported for this resource
+# To import , check import.sh for more info
```
Comment (shenda1) [inline examples/resources/powerflex_snapshot_policy/resource.tf:20]: import.sh is missing

File: powerflex/provider/snapshot_policy_resource_test.go [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+func TestAccSnapshotPolicyResource(t *testing.T) {
+	if os.Getenv("TF_ACC") == "" {
+		t.Skip("Dont run with units tests because it will try to create the context")
+	}
+	//resourceName := "powerflex_snapshot_policy.avengers-fs-create"
+	resource.Test(t, resource.TestCase{
+		ProtoV6ProviderFactories: testAccProtoV6ProviderFactories,
+		Steps: []resource.TestStep{
+			// Create fault set Test
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource_test.go:34]: fault set -> snapshot policy. Applicable for all instances

**Review by shenda1 — CHANGES_REQUESTED** [review]
Please add coverage report

File: examples/resources/powerflex_snapshot_policy/resource.tf (hunk 2 above)
Comment (AnikaAgiwal2711) [reply examples/resources/powerflex_snapshot_policy/resource.tf]: Done

File: powerflex/provider/snapshot_policy_resource_test.go (hunk 3 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource_test.go]: Done

File: powerflex/provider/snapshot_policy_resource.go (hunk 4 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go]: Done

File: powerflex/provider/snapshot_policy_resource.go (hunk 5 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go:157]: Handled the partial creation issue.

File: powerflex/provider/snapshot_policy_resource.go (hunk 7 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go]: Done

File: powerflex/provider/snapshot_policy_resource.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go]: Moved the logic to the update state function

Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go]: Moved the logic to the update state function

File: powerflex/provider/snapshot_policy_schema.go (hunk 9 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_schema.go]: Changed

File: powerflex/provider/snapshot_policy_schema.go (hunk 10 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_schema.go]: Changed

File: examples/resources/powerflex_snapshot_policy/resource.tf (hunk 12 above)
Comment (AnikaAgiwal2711) [reply examples/resources/powerflex_snapshot_policy/resource.tf]: Updated

File: examples/resources/powerflex_snapshot_policy/resource.tf (hunk 13 above)
Comment (AnikaAgiwal2711) [reply examples/resources/powerflex_snapshot_policy/resource.tf:20]: added

**Comment by AnikaAgiwal2711** [comment]
image

File: powerflex/provider/snapshot_policy_resource.go [hunk 15]
```diff
@@ -0,0 +1,464 @@
...
+	// fetching the list of volumes assigned to a snaphot policy
+	volumes, err := r.system.GetSourceVolume(snapID)
+	if err != nil {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Could not get volumes assigned to snapshot policy"),
+			err.Error(),
+		)
+		return
+	}
+	// updating the detils to the state
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource.go:184]: details

File: powerflex/provider/snapshot_policy_resource_test.go [hunk 16]
```diff
@@ -0,0 +1,353 @@
...
+
+var SPResourceCreateWithVol = `
+resource "powerflex_snapshot_policy" "avengers-sp-create" {
+	name = "snap-create-test"
+	num_of_retained_snapshots_per_level = [1]
+	auto_snapshot_creation_cadence_in_min = 5
+	paused = false
+	secure_snapshots = false
+	snapshot_access_mode = "ReadOnly"
+	volume_ids = ["edd2fb3100000007","edd2fb3200000008"]
```
Comment (shenda1) [inline powerflex/provider/snapshot_policy_resource_test.go:198]: Don't use hardcoded values. Applicable for all instances.

**Review by shenda1 — COMMENTED** [review]
@AnikaAgiwal2711 Minor comments else looks good. Is it possible to increase the coverage > 80% ? Also github actions did not run (probably because of the conflict)

**Comment by Krunal-Thakkar** [comment]
Please resolve conflicts

File: powerflex/provider/snapshot_policy_resource_test.go (hunk 16 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource_test.go]: Done

File: powerflex/provider/snapshot_policy_resource.go (hunk 15 above)
Comment (AnikaAgiwal2711) [reply powerflex/provider/snapshot_policy_resource.go]: Done

**Comment by AnikaAgiwal2711** [comment]
> @AnikaAgiwal2711 Minor comments else looks good. Is it possible to increase the coverage > 80% ? Also github actions did not run (probably because of the conflict)

I've added all the corner cases what all was possible. I don't think it will cross 80%

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by Krunal-Thakkar** [comment]
run e2e test

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Comment by AnikaAgiwal2711** [comment]
run e2e test

**Review by shenda1 — APPROVED** [review]
